substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/service/GcNotifier.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2024, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, 2024, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.genscavenge.service;
+
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.Platform;
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.genscavenge.AbstractMemoryPoolMXBean;
+import com.oracle.svm.core.heap.GCCause;
+import com.oracle.svm.core.genscavenge.GenScavengeMemoryPoolMXBeans;
+import com.oracle.svm.core.genscavenge.AbstractGarbageCollectorMXBean;
+import com.oracle.svm.core.thread.VMOperation;
+
+import java.lang.management.GarbageCollectorMXBean;
+import java.lang.management.ManagementFactory;
+
+import static com.oracle.svm.core.genscavenge.GenScavengeMemoryPoolMXBeans.COMPLETE_SCAVENGER;
+import static com.oracle.svm.core.genscavenge.GenScavengeMemoryPoolMXBeans.OLD_GEN_SPACE;
+import static com.oracle.svm.core.genscavenge.GenScavengeMemoryPoolMXBeans.YOUNG_GEN_SCAVENGER;
+
+/**
+ * This class keeps a circular queue of requests to later use when emitting GC notifications. The
+ * requests are pre-allocated so that they can be used during a GC.
+ *
+ * This class is responsible for enqueuing requests during GC safepoints and later dequeuing
+ * requests to generate notifications when the service thread handles signals outside of safepoints.
+ */
+public class GcNotifier {
+    // 5 is probably more than enough, although making the queue larger isn't a problem either.
+    private static final int QUEUE_SIZE = 5;
+    private static volatile int rear;
+    private static volatile int front;
+
+    GcNotificationRequest[] requests;
```

Comment:
I've created a class `CircularQueue` that extends `RingBuffer`, since `RingBuffer` only seems to retain the last N items but doesn't actually track the first/last items added.